### ISSUE TITLE
Model loads after each completed partition

### ISSUE THREAD

**Issue by conceptofmind** [Eventual-Inc/Daft#2878]
Given the embedding udf below, the model re-loads and is reinitializing after each write and completed partition / parquet file:
```python
import daft
import torch
import numpy as np
from typing import Optional

BATCH_SIZE = 1
NUM_GPUS = 1
return_dtype=daft.DataType.list(daft.DataType.float64())

@daft.udf(return_dtype=return_dtype, num_gpus=NUM_GPUS, batch_size=BATCH_SIZE)
class STEmbeddingUDF:
    def __init__(
        self,
        model_name: str,
        device: str = "cuda",
        convert_to_tensor: bool = False,
        torch_dtype: torch.dtype = torch.float16,
        set_seq_len: bool = True,
        max_seq_length: Optional[int] = 2048,
    ):
        from sentence_transformers import SentenceTransformer
        
        self.model = SentenceTransformer(
            model_name,
            device=device,
            model_kwargs={"torch_dtype": torch_dtype}
        )

        if set_seq_len:
            self.model.max_seq_length = max_seq_length

        self.convert_to_tensor = convert_to_tensor
        self.device = device

    def __call__(self, text_col: daft.DataFrame) -> daft.DataFrame:
        embeddings = self.model.encode(
            text_col.to_pylist(), 
            batch_size=BATCH_SIZE,
            convert_to_tensor=self.convert_to_tensor,
            device=self.device,
        )
        if self.convert_to_tensor is not True:
            embeddings = embeddings[0].astype(np.float64) 
        return [embeddings]

...

df = daft.read_parquet(f"hf://datasets/name")

processor = STEmbeddingUDF
data = col("text")
df = df.with_column("embeddings, processor(data))

df.to_parquet("embeddings")
```
Model reloads after completing and writing one partition/file:
```
Loading checkpoint shards: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████| 7/7 [00:49<00:00,  7.09s/it]
We detected that you are passing `past_key_values` as a tuple and this is deprecated and will be removed in v4.43. Please use an appropriate `Cache` class (https://huggingface.co/docs/transformers/v4.41.3/en/internal/generation_utils#transformers.Cache)
Loading checkpoint shards: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 7/7 [00:47<00:00,  6.82s/it]
Loading checkpoint shards:  29%|████████████████████████████████▎                                                                                | 2/7 [00:13<00:32,  6.41s/it]
```
Any input would be greatly appreciated.

**Comment by jaychia** [comment]
Indeed. This is something @kevinzwang is working on stabilizing our solution for. 

You can actually try it out with the environment variable: `DAFT_ENABLE_ACTOR_POOL_PROJECTIONS =1`

The fix is currently available for the local executor. We have some local branches available for the Ray Runner as well and are in the process of testing/stabilizing both. Let us know if you try it!

**Comment by jaychia** [comment]
I see you're using a GPU as well in your UDF. We will probably want to correctly assign the `CUDA_VISBLE_DEVICES` appropriately for each instance of your UDF which isn't yet being done on the PyRunner.

**Comment by conceptofmind** [comment]
> Indeed. This is something @kevinzwang is working on stabilizing our solution for.
> 
> You can actually try it out with the environment variable: `DAFT_ENABLE_ACTOR_POOL_PROJECTIONS =1`
> 
> The fix is currently available for the local executor. We have some local branches available for the Ray Runner as well and are in the process of testing/stabilizing both. Let us know if you try it!

Hi @jaychia ,

I will test out setting the environment variable for cuda visible devices and actor pool.

Thank you!

**Comment by jaychia** [comment]
Made a PR for an initial attempt at doing `CUDA_VISIBLE_DEVICES`: https://github.com/Eventual-Inc/Daft/pull/2882

You'll likely need that if running multi-GPU on a single node + PyRunner!

**Comment by conceptofmind** [comment]
Hi @jaychia,

Setting `DAFT_ENABLE_ACTOR_POOL_PROJECTIONS=1` helped to resolve the issue of the model loading multiple times per completed partition.

But now when I run the above udf with :
```python
STEmbeddingUDF.with_init_args(
    model_name="sentence-transformers/all-MiniLM-L6-v2", 
    convert_to_tensor=False
).with_concurrency(1).override_options(num_gpus=1)
```
I receive the error:
```
    name: partial_udf.func_cls() for name, partial_udf in partial_stateful_udfs.items()
TypeError: STEmbeddingUDF.__init__() missing 1 required positional argument: 'model_name'

    raise self._exception
concurrent.futures.process.BrokenProcessPool: A process in the process pool was terminated abruptly while the future was running or pending.
```
Thank you.

**Comment by jaychia** [comment]
Oops yes, I think you've stumbled upon one of the TODOs in this beta feature:
<img width="699" alt="image" src="https://github.com/user-attachments/assets/c717882f-99db-4454-b376-ded34930b6a0">

Let me work on something here

**Comment by conceptofmind** [comment]
> Oops yes, I think you've stumbled upon one of the TODOs in this beta feature: <img alt="image" width="699" src="https://private-user-images.githubusercontent.com/17691182/369701869-c717882f-99db-4454-b376-ded34930b6a0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjcwMjIwMjQsIm5iZiI6MTcyNzAyMTcyNCwicGF0aCI6Ii8xNzY5MTE4Mi8zNjk3MDE4NjktYzcxNzg4MmYtOTlkYi00NDU0LWIzNzYtZGVkMzQ5MzBiNmEwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTIyVDE2MTUyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZiNjY5OWI1OTQ3ZTE2NWRmZTk2NWUxODkyNzEzZDFiYWRlMWFlMGEyNzI4YjBkNzU5NGEzYjNiZTg1YTZjMTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Es4_0x1PMbdB3dvfUdlVSpWA4cdJKQgtnF6lYm7q-2w">
> 
> Let me work on something here

Thank you!

**Comment by conceptofmind** [comment]
Hi @jaychia, 

I am now able to run with `EmbeddingUDF.with_init_args(model_name).with_concurrency(1).override_options(num_gpus=1)` using the main branch. 

But now hit an error for `RuntimeError: No CUDA GPUs are available` when using `daft.context.set_runner_ray()`.

I have exported `export CUDA_VISIBLE_DEVICES=0` before running the script.

Thank you.

**Comment by jaychia** [comment]
> Hi @jaychia,
> 
> I am now able to run with `EmbeddingUDF.with_init_args(model_name).with_concurrency(1).override_options(num_gpus=1)` using the main branch.
> 
> But now hit an error for `RuntimeError: No CUDA GPUs are available` when using `daft.context.set_runner_ray()`.
> 
> I have exported `export CUDA_VISIBLE_DEVICES=0` before running the script.
> 
> I know you mentioned previously that there was another branch for Ray Actor pool but I was unsure if these were directly related.
> 
> Thank you.

Ahh derp! I think we just forgot to forward the GPU requests to the Actor pool that we're creating. Should be a very simple fix. cc @kevinzwang

**Comment by kevinzwang** [comment]
Hi @conceptofmind, just merged in a PR to fix the ray actor resource allocation issue you encountered. It will be included in our next release!

**Comment by conceptofmind** [comment]
> Hi @conceptofmind, just merged in a PR to fix the ray actor resource allocation issue you encountered. It will be included in our next release!

Hi @kevinzwang,

I will test it out on the main branch now.

Thank you!

**Comment by conceptofmind** [comment]
Looks like it is all working now on the main branch.

See 1.0/1.0 in the Ray Actor pool:
```
1.0/1.0 GPU
```
I will look into testing on multiple GPUs soon.

**Comment by kevinzwang** [comment]
Great to hear! By the way, feel free to close this issue once you confirm that this problem is resolved.

**Comment by conceptofmind** [comment]
Yes will close out after testing multi gpu single node and multi GPU multi node.

**Comment by kevinzwang** [comment]
Closing this issue, but feel free to reopen if the fix did not work
